ansible/roles-infra/infra-osp-resources-destroy/tasks/project.yml

```diff
@@ -2,9 +2,38 @@
     OS_AUTH_URL: "{{ osp_auth_url }}"
     OS_USERNAME: "{{ osp_auth_username }}"
     OS_PASSWORD: "{{ osp_auth_password }}"
+    OS_PROJECT_NAME: "{{ osp_project_name }}"
     OS_PROJECT_DOMAIN_ID: "{{ osp_auth_project_domain }}"
     OS_USER_DOMAIN_NAME: "{{ osp_auth_user_domain }}"
   block:
+    - name: Get user info
+      os_user_facts:
+        name: "{{ guid }}-user"
+        domain: default
+      register: r_osp_user_info
+
+    - name: Get UUID of user
+      set_fact:
+        osp_user_uuid: "{{ r_osp_user_info | json_query(uuid_query) }}"
+      vars:
+        uuid_query: ansible_facts.openstack_users[].id|[0]
+
+    # These can't use the os_ module since impersonating the user is only
+    # available with the nova client here.
+    - name: Get keypairs for user
+      shell: nova keypair-list --user {{ osp_user_uuid }} | grep ssh | awk '{ print $2 }'
+      register: r_user_keypairs
+
+    - name: Delete user keypair
+      when: r_user_keypairs.stdout_lines | length > 0
+      shell: nova keypair-delete --user {{ osp_user_uuid }} {{ item }}
+      loop: "{{ r_user_keypairs.stdout_lines }}"
```

Comment:
Why do we delete keypairs as admin here?

If they are part of the user Heat stack, when we delete the stack, they should be deleted there?

Reply:
We do not know the password for the user during teardown so I cannot do things as the user. We have to delete things as admin, that is the only credential we know. This allows us to delete the keypair "on behalf" of the user.

Reply:
OK but when the admin deletes the user stack, the keys are gone, no?